### DISCUSSION TITLE
Raspberry Pi Zero-powered Mass Storage Device

### DISCUSSION THREAD

**Discussion by ejona86** [keirf/flashfloppy#451] [Ideas]
The Raspberry Pi Zero supports USB-OTG and USB gadgets on Linux are pretty trivial to set up. This provides some interesting possibilities for $5-$10. Any device that can act as a USB device could work, but the Zero is inexpensive and accessible.

With a 2-line configuration change you can take a normal Pi distribution (I used Arch Linux ARM, but Raspbian would work as well) and have it expose a file as a USB Mass Storage Device (so the Pi would act like a thumb drive). I added `dtoverlay=dwc2` to boot/config.txt and `g_mass_storage.file=/root/zerodisk.img modules-load=dwc2,g_mass_storage` to boot/cmdline.txt. Most of the effort is to create zerodisk.img and put your floppy images inside it. You might also want `dtoverlay=gpio-shutdown` in config.txt to get a shutdown button; read overlays/README for details.

With such a set up, writes can be faster than reads because there is no need to wait for backing storage. The entire floppy disk can trivially be buffered in 1 GB of RAM so we'd mainly be limited by USB Full Speed data rate (12 Mbps). Scheduling jitter from Linux doesn't seem to be a problem with some limited testing. 512 byte writes are 2.5-4µs. 2k reads are 6-10µs.

(Note: The Zero has two USB ports, one for power and one for power+data. However, you must not power the device from both ports at the same time as there is no protection circuit. This means that without some special preparations it has to be shut down and booted again if you change which machine it is plugged in to.)

That is "good enough" for me as a solution in case USB drive write latencies cause me trouble. But I sort of expect others might want to take it further.

*Wifi*. Since the Zero W supports Wifi, you can SSH to it, run `rmmod g_mass_storage` which "unplugs" the "flash drive", change the "flash drive" contents in the img file, and then `modprobe g_mass_storage` to "plug it back in". This is enough to allow a web-based GUI for choosing the floppy image.

*USB serial*. There is a serial gadget (the zero acts like a USB serial adapter). This would need some investigation to determine the effort to support the serial USB protocol in FlashFloppy, but there's the ability here to have rich communicating between the two systems.

**Comment by ejona86** [comment]
Doh. This was _just recently_ brought up in https://github.com/keirf/FlashFloppy/issues/180#issuecomment-792954388 . I didn't see it until now! I had seen the issue, but it all looked old and generic "serial API" discussion. (I actually started down this path via the Raspberry Pi Pico and stm32 Black Pill which I had noticed supported USB-OTG.)

**Reply by keirf** [reply]
Yes this furrow has been ploughed before. But the insight about the benefits of the PiZ's RAM as buffer cache is new. It sounds a very nice arrangement for your hard-sectored scenario. I still have it on the backburner to port to a buildroot-based Linux board like PiZ or OPiZ. But it's way more work and this gets most of the benefit.

**Reply by peter-mount** [reply]
> Doh. This was _just recently_ brought up in [#180 (comment)](https://github.com/keirf/FlashFloppy/issues/180#issuecomment-792954388) . I didn't see it until now! I had seen the issue, but it all looked old and generic "serial API" discussion. (I actually started down this path via the Raspberry Pi Pico and stm32 Black Pill which I had noticed supported USB-OTG.)

Yes that was me. Here's my blog from last year from when I did this on a [BBC Master](https://area-51.blog/2020/01/30/using-a-raspberry-pi-zero-w-to-hold-disk-images-for-the-bbc-micro/). It's mostly errata for an earlier Raspberry PI blog article on how to setup a PI 0 but their CMS has broken the formatting.

It works well & it's not a file exposed to the GoTek but a Fat formatted partition so both can access it at the same time (with some limits) but probably could be updated if things have changed over the last 14 months.

I use PI 0W's for this sort of thing as I can then SCP new disk images to the machine easily, actually did that last night for a [Twitter thread](https://twitter.com/peter_mount/status/1373392644219473924) when looking at an issue with the BBCMicroBot & illegal 6502 instructions. Created a disk, copied it over to the PI then ran the disk.
